public/app/features/notifications/StoredNotifications.tsx

```diff
@@ -17,19 +17,38 @@ import { useDispatch, useSelector } from 'app/types';
 export function StoredNotifications() {
   const dispatch = useDispatch();
   const notifications = useSelector((state) => selectWarningsAndErrors(state.appNotifications));
+  const [selectedNotificationIds, setSelectedNotificationIds] = useState<string[]>([]);
+  const allNotificationsSelected = notifications.every((notification) =>
+    selectedNotificationIds.includes(notification.id)
+  );
   const lastReadTimestamp = useRef(useSelector((state) => selectLastReadTimestamp(state.appNotifications)));
   const styles = useStyles2(getStyles);
 
   useEffectOnce(() => {
     dispatch(readAllNotifications(Date.now()));
   });
 
-  const onClearNotification = (id: string) => {
-    dispatch(clearNotification(id));
+  const clearSelectedNotifications = () => {
+    if (allNotificationsSelected) {
+      dispatch(clearAllNotifications());
+    } else {
+      selectedNotificationIds.forEach((id) => {
+        dispatch(clearNotification(id));
+      });
+    }
+    setSelectedNotificationIds([]);
+  };
+
+  const handleAllCheckboxToggle = (isChecked: boolean) => {
+    setSelectedNotificationIds(isChecked ? notifications.map((n) => n.id) : []);
   };
 
-  const clearAllNotifs = () => {
-    dispatch(clearAllNotifications());
+  const handleCheckboxToggle = (id: string, isChecked: boolean) => {
+    if (isChecked && !selectedNotificationIds.includes(id)) {
+      setSelectedNotificationIds([...selectedNotificationIds, id]);
+    } else if (!isChecked && selectedNotificationIds.includes(id)) {
+      setSelectedNotificationIds(selectedNotificationIds.filter((notificationId) => notificationId !== id));
+    }
```

Comment:
Because you're operating on the value of the state, use the callback form of `setState`:

```ts
setSelectedNotificationIds(currentValue => {
    if (isChecked && !currentValue .includes(id)) {
      return [...currentValue , id];
    } else if (!isChecked && currentValue .includes(id)) {
      return currentValue .filter((notificationId) => notificationId !== id)
    }
})
```